### ISSUE TITLE
feat: Implement Tenstorrent Backend Support (TT-NN)

### ISSUE THREAD

**Issue by cognitive-glitch** [cognitive-glitch/CTranslate2-TT#2]
### **Description**

This proposal outlines the architectural plan to integrate **Tenstorrent** hardware support (Grayskull, Wormhole) into CTranslate2. The implementation will utilize the **TT-NN (Tenstorrent Neural Network)** SDK to provide a high-performance backend, enabling CTranslate2 to execute large-scale Transformer inference (Llama, Whisper, BERT) on Tenstorrent AI accelerators.

### **Motivation**

CTranslate2 is a standard-setting inference engine for Transformer models. Extending support to Tenstorrent hardware expands the ecosystem to RISC-V-based AI accelerators, offering high throughput and energy efficiency while maintaining the engine's lightweight, framework-agnostic nature.

#### **1. The Memory "Impedance Mismatch"**

* **Challenge:** CTranslate2 uses raw pointers (`void*`) for memory handles and relies on pointer arithmetic (e.g., `data + offset`) for slicing. The Tenstorrent `ttnn` library manages memory via opaque `ttnn::Tensor` objects, which encapsulate device buffer handles, shape, and SRAM banking configuration.
* **Solution: The Address-Based Registry**
* We will implement a custom `TTAllocator` that acts as a translation layer.
* **Allocation:** The allocator creates a `ttnn::Tensor` on the device, extracts its persistent DRAM buffer address, registers the mapping `Address -> Tensor` in a global registry, and returns the address as the `void*` handle.
* **Resolution:** When a primitive receives a `void* ptr`, it queries the registry. If `ptr` is an offset (e.g., slicing), the registry logic finds the parent tensor and calculates the offset relative to the base address.

#### **2. Layout Management**

* **Challenge:** TT hardware compute units operate on 32x32 tiles (`TILE_LAYOUT`) for optimal performance, whereas CTranslate2 assumes `ROW_MAJOR_LAYOUT` on the host. Runtime conversion is expensive.
* **Solution:**
* **Implicit Tilization:** Data transfer from Host -> Device (`copy_from`) will implicitly convert supported shapes to `TILE_LAYOUT`.
* **Persistent Layout:** Intermediate activations (output of MatMuls, Add, Gelu) remain in `TILE_LAYOUT` on the device.
* **Lazy Untilization:** Data is only converted back to `ROW_MAJOR` when explicitly copied back to the Host (`copy_to`).
* **Weight Pre-Packing:** Linear layer weights are converted to `TILE_LAYOUT` and `BFLOAT16` (or `BFLOAT8_B`) during the `Model::load` phase.

---

#### **Phase 1: Build System & Environment**

**Goal:** Configure CMake to detect and link the Tenstorrent SDK.

* [ ] **1.1 Docker Support**
* Create `docker/Dockerfile.tt` based on the official `tenstorrent/tt-metalium` images.

* [ ] **1.2 CMake Configuration**
* Add `option(WITH_TENSTORRENT "Compile with Tenstorrent backend" OFF)` in `CMakeLists.txt`.
* Create `cmake/FindTenstorrent.cmake` to locate `libttnn.so`, `libtt_metal.so`, and headers using standard env vars (`TT_METAL_HOME`).
* Define preprocessor definition `CT2_WITH_TENSTORRENT`.

#### **Phase 2: Core Runtime & Device Management**

**Goal:** Enable the engine to recognize the `TT` device and manage the connection context.

* [ ] **2.1 Device Integration**
* Update `include/ctranslate2/devices.h` to add `Device::TT`.
* Update `src/devices.cc` (`str_to_device`, `device_to_str`) to handle `"tt"` strings.

* [ ] **2.2 Backend Context**
* Create `src/tt/backend.h` and `src/tt/backend.cc`.
* Implement a Singleton `TTContext` responsible for calling `ttnn::device::open_device(id)` on startup and `close_device` on shutdown.

#### **Phase 3: Memory Management (The Registry)**

**Goal:** Implement the translation layer between `void*` and `ttnn::Tensor`.

* [ ] **3.1 TT Allocator**
* Create `src/tt/allocator.cc` inheriting from `Allocator`.
* **Registry:** Maintain a thread-safe `std::map<uint64_t, ttnn::Tensor> _registry`.
* **Allocate:** Call `ttnn::empty`, register `buffer->address()`, return address.
* **Free:** Look up address, erase from registry (triggering Tensor destructor).

* [ ] **3.2 Tensor Resolution Utilities**
* Create `src/tt/utils.h`.
* Implement `ttnn::Tensor resolve_tensor(void* ptr, Shape expected_shape)`.
* **Logic:** Use `_registry.upper_bound(ptr)` to find the allocation block covering the pointer. Calculate `offset = ptr - base_addr`. Return the tensor (applying a slice if `offset > 0`).

#### **Phase 4: Data Transfer & StorageView**

**Goal:** Handle data movement and layout conversion.

* [ ] **4.1 StorageView Modifications**
* Modify `src/storage_view.cc` to handle `Device::TT` dispatch.

* [ ] **4.2 Copy Host -> TT**
* Resolve registry entry.
* Create a host-side `ttnn::Tensor` from the source buffer.
* Use `ttnn::to_device`. If shape is tile-compatible (dim % 32 == 0), use `Layout::TILE`.

* [ ] **4.3 Copy TT -> Host**
* Resolve registry entry.
* **Check Layout:** If `TILE_LAYOUT`, run `ttnn::to_layout(..., ROW_MAJOR)` before reading.
* Read back to host buffer.

#### **Phase 5: Primitives & Operators**

**Goal:** Implement the math kernels using TT-NN.

* [ ] **5.1 Primitive Infrastructure**
* Create `src/tt/primitives.cc`.
* Specialize `primitives<Device::TT>` templates.

* [ ] **5.2 Basic Ops**
* `add`, `sub`, `mul`, `div` (Broadcast support required).
* `relu`, `gelu`, `tanh`, `sigmoid`.
* `layer_norm`, `softmax` (Map directly to `ttnn` optimized kernels).

* [ ] **5.3 Matrix Multiplication (GEMM)**
* Implement `Gemm` primitive.
* **Logic:** Check inputs. If Row-Major, `ttnn::to_layout(..., TILE)`. Call `ttnn::matmul`.
* **Output:** Store the result in the registry. Keep output in `TILE_LAYOUT` (crucial optimization).

#### **Phase 6: Advanced Optimizations**

**Goal:** Performance tuning for Transformer workloads.

* [ ] **6.1 Weight Pre-Packing**
* Modify `src/models/model.cc`.
* When loading for `Device::TT`, detect Linear weights.
* Convert to `TILE_LAYOUT` and `BFLOAT16` *immediately* upon load.

* [ ] **6.2 Flash Attention**
* Create `src/ops/flash_attention_tt.cc`.
* Bind `FlashAttention` op to `ttnn::transformer::scaled_dot_product_attention`.

---

### **Technical Deep Dive: The Registry Logic**

Because CTranslate2 passes `void*` pointers that may represent an offset into a larger buffer, the backend must resolve the owning tensor dynamically.

```cpp
// src/tt/utils.cc Concept
ttnn::Tensor resolve_tensor(const void* ptr) {
    uint64_t addr = (uint64_t)ptr;
    std::lock_guard<std::mutex> lock(registry_mutex);
    
    // Find the first key strictly greater than addr, then decrement iterator
    // to find the allocation block that starts at or before addr.
    auto it = _registry.upper_bound(addr);
    
    if (it != _registry.begin()) {
        --it;
        uint64_t base_addr = it->first;
        ttnn::Tensor& base_tensor = it->second;
        size_t offset = addr - base_addr;
        
        // Validation: Ensure offset is within tensor bounds
        // ...
        
        if (offset == 0) return base_tensor;
        
        // Return a slice or view starting at offset
        // Note: TTNN slicing requires care regarding Tile alignment
        return ttnn::slice(base_tensor, offset_to_shape(offset));
    }
    throw std::runtime_error("Pointer not found in TT Registry");
}

```

### **Risks & Unknowns**

* **Async Synchronization:** CTranslate2 assumes asynchronous execution in places. We must ensure `synchronize_device` correctly flushes the TT command queue.
* **Driver Compatibility:** Ensuring the linked `libttnn.so` matches the installed host driver version.
* **Memory Fragmentation:** Heavy use of `empty` allocations without a pooling allocator might lead to fragmentation on device DRAM.

### **Acceptance Criteria**

1. **Builds**: Successfully compiles with `-DWITH_TENSTORRENT=ON` inside the Tenstorrent Docker container.
2. **Tests**: Passes `tests/ctranslate2_test` for `Device::TT` (focusing on Gemm, Add, Softmax).
3. **End-to-End**: A standard model (e.g., Llama-2-7b) runs on Grayskull/Wormhole and produces output logits matching CPU execution.
